File: DataReaderToPSObject.vb
Public Class DataReaderToPSObject
    Private Class Map
        Public Ordinal As Integer
        Public Name As String
        'Public AllowNull As Boolean
        Public DataType As String

        Public Sub New(ordinal As Integer, datatype As String, Optional name As String = Nothing) 'allownull As Boolean, 
            If String.IsNullOrWhiteSpace(name) Then name = String.Format("Column{0}", ordinal + 1)

            Me.Ordinal = ordinal
            Me.Name = name
            'Me.AllowNull = allownull
            Me.DataType = datatype
        End Sub

    End Class

    Public Shared Iterator Function Translate(dataReader As IDataReader) As IEnumerable(Of PSObject)
        Dim MapList As New List(Of Map), Ord As Integer = 0
        For Each x In dataReader.GetSchemaTable.AsEnumerable.OrderBy(Function(r) r("ColumnOrdinal")).ToList
            MapList.Add(New Map(Ord, x("DataType").Name, x("ColumnName"))) 'x("AllowDBNull"),
            Ord += 1
        Next


        While dataReader.Read
            Dim psObj As New PSObject
            For Each m As Map In MapList
                With psObj.Members
                    If dataReader.IsDBNull(m.Ordinal) Then 'm.AllowNull AndAlso dataReader.IsDBNull(m.Ordinal) Then 
                        .Add(New PSNoteProperty(m.Name, Nothing), True)
                    Else
                        Try
                            Select Case m.DataType
                                Case "Boolean"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetBoolean(m.Ordinal)), True)
                                Case "Byte"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetByte(m.Ordinal)), True)
                                Case "Char"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetChar(m.Ordinal)), True)
                                Case "DateTime"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetDateTime(m.Ordinal)), True)
                                Case "Decimal"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetDecimal(m.Ordinal)), True)
                                Case "Double"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetDouble(m.Ordinal)), True)
                                Case "Single"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetFloat(m.Ordinal)), True)
                                Case "Guid"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetGuid(m.Ordinal)), True)
                                Case "Int16"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetInt16(m.Ordinal)), True)
                                Case "Int32"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetInt32(m.Ordinal)), True)
                                Case "Int64"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetInt64(m.Ordinal)), True)
                                Case "String"
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetString(m.Ordinal)), True)
                                Case Else
                                    .Add(New PSNoteProperty(m.Name, dataReader.GetValue(m.Ordinal)), True)
                            End Select
                        Catch ex As Exception
                            Dim msg As String = String.Format("Failed to translate, ColumnName = {0} | ColumnOrdinal = {1} | ColumnType = {2} | ToStringValue = '{3}' | See InnerException for details", m.Name, m.Ordinal, m.DataType, dataReader.GetValue(m.Ordinal).ToString)
                            Throw New Exception(msg, ex)
                        End Try
                    End If
                End With
            Next
            Yield psObj
        End While

    End Function
End Class

File: DataReaderToPSObject/DataReaderToPSObject.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Management.Automation;

public class DataReaderToPSObject
{
    private class Map
    {
        public int Ordinal;
        public string Name;
        // Public AllowNull As Boolean
        public string DataType;

        public Map(int ordinal, string datatype, string name = null) // allownull As Boolean, 
        {
            if (string.IsNullOrWhiteSpace(name))
                name = string.Format("Column{0}", ordinal + 1);

            this.Ordinal = ordinal;
            this.Name = name;
            // Me.AllowNull = allownull
            this.DataType = datatype;
        }
    }

    public static IEnumerable<PSObject> Translate(System.Data.Common.DbDataReader dataReader, Boolean ProviderTypes)
    {
        List<Map> MapList = new List<Map>();
        int Ord = 0;
        foreach (var x in dataReader.GetSchemaTable().Select("", "ColumnOrdinal"))
        {
            MapList.Add(new Map(Ord, x["DataType"].ToString(), x["ColumnName"].ToString())); // x("AllowDBNull"),
            Ord += 1;
        }

        PSObject responseObject = new PSObject();
        while (dataReader.Read())
        {
            PSObject psObj = new PSObject();
            foreach (Map m in MapList)
            {
                {
                    var withBlock = psObj.Members;
                    if (dataReader.IsDBNull(m.Ordinal))
                        withBlock.Add(new PSNoteProperty(m.Name, null), true);
                    else
                        try
                        {
                            if (ProviderTypes)
                                withBlock.Add(new PSNoteProperty(m.Name, dataReader.GetProviderSpecificValue(m.Ordinal)), true);
                            else
                                withBlock.Add(new PSNoteProperty(m.Name, dataReader.GetValue(m.Ordinal)), true);
                        }
                        catch (Exception ex)
                        {
                            string msg = string.Format("Failed to translate, ColumnName = {0} | ColumnOrdinal = {1} | ColumnType = {2} | ToStringValue = '{3}' | See InnerException for details", m.Name, m.Ordinal, m.DataType, dataReader.GetValue(m.Ordinal).ToString());
                            throw new Exception(msg, ex);
                        }
                }
            }
            yield return psObj;
        }
    }
}